diff --git a/src/Lean/Linter/UnusedSimpArgs.lean b/src/Lean/Linter/UnusedSimpArgs.lean
index 1c8f04bef6..a620b76ed3 100644
--- a/src/Lean/Linter/UnusedSimpArgs.lean
+++ b/src/Lean/Linter/UnusedSimpArgs.lean
@@ -29,7 +29,7 @@ private def warnUnused (stx : Syntax) (i : Nat) : CoreM Unit := do
     otherArgs := otherArgs.push simpArgs[j]
   let stx' := Tactic.setSimpParams stx otherArgs
   let suggestion : Meta.Hint.Suggestion := stx'
-  let suggestion := { suggestion with span? := stx }
+  let suggestion := { suggestion with span? := stx, diffGranularity := .none }
   let mut hint ← MessageData.hint "Omit it from the simp argument list." #[ suggestion ]
   -- Add hint about `←`
   let isInv := argStx.getKind == ``Lean.Parser.Tactic.simpLemma && !argStx[1].isNone
