error[E0277]: the size for values of type `T` cannot be known at compilation time
 --> tests/ui/pin_project/project_replace_unsized.rs:6:8
  |
6 | struct Struct<T: ?Sized> {
  |        ^^^^^^^-^^^^^^^^^
  |        |      |
  |        |      this type parameter needs to be `Sized`
  |        doesn't have a size known at compile-time
  |
note: required because it appears within the type `__StructProjectionOwned<T>`
 --> tests/ui/pin_project/project_replace_unsized.rs:6:8
  |
6 | struct Struct<T: ?Sized> {
  |        ^^^^^^
  = note: the return type of a function must have a statically known size
help: consider removing the `?Sized` bound to make the type parameter `Sized`
  |
6 - struct Struct<T: ?Sized> {
6 + struct Struct<T> {
  |

error[E0277]: the size for values of type `T` cannot be known at compilation time
  --> tests/ui/pin_project/project_replace_unsized.rs:11:8
   |
11 | struct TupleStruct<T: ?Sized>(T);
   |        ^^^^^^^^^^^^-^^^^^^^^^
   |        |           |
   |        |           this type parameter needs to be `Sized`
   |        doesn't have a size known at compile-time
   |
note: required because it appears within the type `__TupleStructProjectionOwned<T>`
  --> tests/ui/pin_project/project_replace_unsized.rs:11:8
   |
11 | struct TupleStruct<T: ?Sized>(T);
   |        ^^^^^^^^^^^
   = note: the return type of a function must have a statically known size
help: consider removing the `?Sized` bound to make the type parameter `Sized`
   |
11 - struct TupleStruct<T: ?Sized>(T);
11 + struct TupleStruct<T>(T);
   |

error[E0277]: the size for values of type `T` cannot be known at compilation time
 --> tests/ui/pin_project/project_replace_unsized.rs:5:15
  |
5 | #[pin_project(project_replace)] //~ ERROR E0277
  |               ^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
6 | struct Struct<T: ?Sized> {
  |               - this type parameter needs to be `Sized`
  |
note: required because it appears within the type `Struct<T>`
 --> tests/ui/pin_project/project_replace_unsized.rs:6:8
  |
6 | struct Struct<T: ?Sized> {
  |        ^^^^^^
  = help: unsized fn params are gated as an unstable feature
help: consider removing the `?Sized` bound to make the type parameter `Sized`
  |
6 - struct Struct<T: ?Sized> {
6 + struct Struct<T> {
  |
help: function arguments must have a statically known size, borrowed types always have a known size
  |
5 | #[pin_project(&project_replace)] //~ ERROR E0277
  |               +

error[E0277]: the size for values of type `T` cannot be known at compilation time
 --> tests/ui/pin_project/project_replace_unsized.rs:5:1
  |
5 | #[pin_project(project_replace)] //~ ERROR E0277
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
6 | struct Struct<T: ?Sized> {
  |               - this type parameter needs to be `Sized`
  |
note: required because it appears within the type `Struct<T>`
 --> tests/ui/pin_project/project_replace_unsized.rs:6:8
  |
6 | struct Struct<T: ?Sized> {
  |        ^^^^^^
note: required by an implicit `Sized` bound in `UnsafeOverwriteGuard`
 --> src/lib.rs
  |
  |     pub struct UnsafeOverwriteGuard<T> {
  |                                     ^ required by the implicit `Sized` requirement on this type parameter in `UnsafeOverwriteGuard`
  = note: this error originates in the derive macro `::pin_project::__private::__PinProjectInternalDerive` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider removing the `?Sized` bound to make the type parameter `Sized`
  |
6   - struct Struct<T: ?Sized> {
6   + struct Struct<T> {
    |

error[E0277]: the size for values of type `T` cannot be known at compilation time
 --> tests/ui/pin_project/project_replace_unsized.rs:7:5
  |
5 | #[pin_project(project_replace)] //~ ERROR E0277
  | ------------------------------- required by a bound introduced by this call
6 | struct Struct<T: ?Sized> {
  |               - this type parameter needs to be `Sized`
7 |     f: T,
  |     ^ doesn't have a size known at compile-time
  |
note: required by an implicit `Sized` bound in `std::ptr::read`
 --> $RUST/core/src/ptr/mod.rs
  |
  | pub const unsafe fn read<T>(src: *const T) -> T {
  |                          ^ required by the implicit `Sized` requirement on this type parameter in `read`
help: consider removing the `?Sized` bound to make the type parameter `Sized`
  |
6    - struct Struct<T: ?Sized> {
6    + struct Struct<T> {
     |

error[E0277]: the size for values of type `T` cannot be known at compilation time
  --> tests/ui/pin_project/project_replace_unsized.rs:10:15
   |
10 | #[pin_project(project_replace)] //~ ERROR E0277
   |               ^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
11 | struct TupleStruct<T: ?Sized>(T);
   |                    - this type parameter needs to be `Sized`
   |
note: required because it appears within the type `TupleStruct<T>`
  --> tests/ui/pin_project/project_replace_unsized.rs:11:8
   |
11 | struct TupleStruct<T: ?Sized>(T);
   |        ^^^^^^^^^^^
   = help: unsized fn params are gated as an unstable feature
help: consider removing the `?Sized` bound to make the type parameter `Sized`
   |
11 - struct TupleStruct<T: ?Sized>(T);
11 + struct TupleStruct<T>(T);
   |
help: function arguments must have a statically known size, borrowed types always have a known size
   |
10 | #[pin_project(&project_replace)] //~ ERROR E0277
   |               +

error[E0277]: the size for values of type `T` cannot be known at compilation time
  --> tests/ui/pin_project/project_replace_unsized.rs:10:1
   |
10 | #[pin_project(project_replace)] //~ ERROR E0277
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
11 | struct TupleStruct<T: ?Sized>(T);
   |                    - this type parameter needs to be `Sized`
   |
note: required because it appears within the type `TupleStruct<T>`
  --> tests/ui/pin_project/project_replace_unsized.rs:11:8
   |
11 | struct TupleStruct<T: ?Sized>(T);
   |        ^^^^^^^^^^^
note: required by an implicit `Sized` bound in `UnsafeOverwriteGuard`
  --> src/lib.rs
   |
   |     pub struct UnsafeOverwriteGuard<T> {
   |                                     ^ required by the implicit `Sized` requirement on this type parameter in `UnsafeOverwriteGuard`
   = note: this error originates in the derive macro `::pin_project::__private::__PinProjectInternalDerive` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider removing the `?Sized` bound to make the type parameter `Sized`
   |
11  - struct TupleStruct<T: ?Sized>(T);
11  + struct TupleStruct<T>(T);
    |
