--- v4.30.0-rc2/Lean/Data/PersistentArray.lean
+++ v4.34.1/Lean/Data/PersistentArray.lean
@@ -227,7 +227,7 @@
 set_option linter.unusedVariables.funArgs false in
 @[specialize]
 partial def forInAux {α : Type u} {β : Type v} {m : Type v → Type w} [Monad m] [inh : Inhabited β]
-    (f : α → β → m (ForInStep β)) (n : PersistentArrayNode α) (b : β) : m (ForInStep β) := do
+    (f : α → β → m (ForInStep β)) (n : @&PersistentArrayNode α) (b : β) : m (ForInStep β) := do
   let mut b := b
   match n with
   | leaf vs =>
@@ -243,7 +243,7 @@
       | ForInStep.yield bNew => b := bNew
     return ForInStep.yield b
 
-@[specialize] protected def forIn (t : PersistentArray α) (init : β) (f : α → β → m (ForInStep β)) : m β := do
+@[specialize] protected def forIn (t : @&PersistentArray α) (init : β) (f : α → β → m (ForInStep β)) : m β := do
   match (← forInAux (inh := ⟨init⟩) f t.root init) with
   | ForInStep.done b  => pure b
   | ForInStep.yield b =>
@@ -397,10 +397,6 @@
 @[inline] def mkPArray {α : Type u} (n : Nat) (v : α) : PArray α :=
   mkPersistentArray n v
 
-end Lean
-
-open Lean (PersistentArray)
-
 /--
 Converts a list to a persistent array.
 -/
@@ -412,3 +408,5 @@
 
 def Array.toPArray' {α : Type u} (xs : Array α) : PersistentArray α :=
   xs.foldl (init := .empty) fun p x => p.push x
+
+end Lean
--- v4.30.0-rc2/Lean/Data/PersistentHashMap.lean
+++ v4.34.1/Lean/Data/PersistentHashMap.lean
@@ -12,6 +12,7 @@
 public import Init.Data.Array.Basic
 import Init.Data.String.Defs
 import Init.Data.ToString.Macro
+import Init.Data.Array.Lemmas
 
 public section
 
@@ -153,7 +154,7 @@
     else findAtAux keys vals heq (i+1) k
   else none
 
-partial def findAux [BEq α] : Node α β → USize → α → Option β
+partial def findAux [BEq α] : @&Node α β → USize → α → Option β
   | Node.entries entries, h, k =>
     let j     := (mod2Shift h shift).toNat
     match entries[j]! with
@@ -162,7 +163,7 @@
     | Entry.entry k' v => if k == k' then some v else none
   | Node.collision keys vals heq, _, k => findAtAux keys vals heq 0 k
 
-def find? {_ : BEq α} {_ : Hashable α} : PersistentHashMap α β → α → Option β
+def find? {_ : BEq α} {_ : Hashable α} : @&PersistentHashMap α β → α → Option β
   | { root }, k => findAux root (hash k |>.toUSize) k
 
 instance {_ : BEq α} {_ : Hashable α} : GetElem (PersistentHashMap α β) α (Option β) fun _ _ => True where
@@ -184,7 +185,7 @@
     else findEntryAtAux keys vals heq (i+1) k
   else none
 
-partial def findEntryAux [BEq α] : Node α β → USize → α → Option (α × β)
+partial def findEntryAux [BEq α] : @&Node α β → USize → α → Option (α × β)
   | Node.entries entries, h, k =>
     let j     := (mod2Shift h shift).toNat
     match entries[j]! with
@@ -193,7 +194,7 @@
     | Entry.entry k' v => if k == k' then some (k', v) else none
   | Node.collision keys vals heq, _, k => findEntryAtAux keys vals heq 0 k
 
-def findEntry? {_ : BEq α} {_ : Hashable α} : PersistentHashMap α β → α → Option (α × β)
+def findEntry? {_ : BEq α} {_ : Hashable α} : @&PersistentHashMap α β → α → Option (α × β)
   | { root }, k => findEntryAux root (hash k |>.toUSize) k
 
 partial def findKeyDAtAux [BEq α] (keys : Array α) (vals : Array β) (heq : keys.size = vals.size) (i : Nat) (k : α) (k₀ : α) : α :=
@@ -288,6 +289,60 @@
     let h := hash k |>.toUSize
     { root := eraseAux root h k }
 
+@[specialize]
+partial def alterAux [BEq α] [Hashable α] (f : Option β → Option β) : Node α β → USize →
+    USize → α → Node α β
+  | n@(Node.collision keys vals heq), h, d, k =>
+    match keys.finIdxOf? k with
+    | some idx =>
+      have : idx < vals.size := heq ▸ idx.isLt
+      let v' := vals[idx]
+      /-
+      This could also be done with more `unsafe` to avoid erase then push but is hopefully not the
+      bottleneck.
+      -/
+      let keys := keys.eraseIdx idx
+      let vals := vals.eraseIdx idx
+      match f (some v') with
+      | some v =>
+        let keys := keys.push k
+        let vals := vals.push v
+        Node.collision keys vals (by simp +zetaDelta [heq])
+      | none   =>
+        Node.collision keys vals (by simp +zetaDelta [heq])
+    | none     =>
+      match f none with
+      | some v => insertAux n h d k v
+      | none   => n
+  | Node.entries entries, h, d, k =>
+    let j     := (mod2Shift h shift).toNat
+    Node.entries <| entries.modify j fun
+      | Entry.null        =>
+        match f none with
+        | some v => Entry.entry k v
+        | none   => Entry.null
+      | Entry.entry k' v' =>
+        if k == k' then
+          match f (some v') with
+          | some v => Entry.entry k v
+          | none   => Entry.null
+        else
+          match f none with
+          | some v => Entry.ref <| mkCollisionNode k' v' k v
+          | none => Entry.entry k' v'
+      | Entry.ref node    =>
+        let newNode := alterAux f node (div2Shift h shift) (d + 1) k
+        match isUnaryNode newNode with
+        | none        => Entry.ref newNode
+        | some (k, v) => Entry.entry k v
+
+@[inline]
+def alter {_ : BEq α} {_ : Hashable α} : PersistentHashMap α β → α → (Option β → Option β) →
+    PersistentHashMap α β
+  | { root }, k, f =>
+    let h := hash k |>.toUSize
+    { root := alterAux f root h 1 k }
+
 section
 variable {m : Type w → Type w'} [Monad m]
 variable {σ : Type w}
@@ -320,7 +375,7 @@
   Id.run <| map.foldlM (pure <| f · · ·) init
 
 protected def forIn {_ : BEq α} {_ : Hashable α} [Monad m]
-    (map : PersistentHashMap α β) (init : σ) (f : α × β → σ → m (ForInStep σ)) : m σ := do
+    (map : @&PersistentHashMap α β) (init : σ) (f : α × β → σ → m (ForInStep σ)) : m σ := do
   let intoError : ForInStep σ → Except σ σ
   | .done s => .error s
   | .yield s => .ok s
